service/src/IOUring.cpp

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include "config.h"
+
+#include "IOUring.hpp"
+
+#include "IOUringFallback.hpp"
+
+#include <iostream>
+#ifdef GEOPM_HAS_IO_URING
+#include "IOUringImp.hpp"
+#endif
+
+namespace geopm
+{
+#ifdef GEOPM_HAS_IO_URING
+    static void emit_missing_support_warning()
+    {
+#ifdef GEOPM_DEBUG
+        static bool do_emit_warning = true;
+        if (do_emit_warning) {
+            do_emit_warning = false;
+            std::cerr << "Warning: <geopm> GEOPM was built with liburing "
+                         "enabled, but the system does not support all uring operations "
+                         "needed by GEOPM. Using non-uring IO instead."
+                      << std::endl;
+        }
+#endif
+    }
+#endif
+
+    std::shared_ptr<IOUring> make_io_uring(unsigned entries)
```

Comment:
I think a unique_ptr return type is preferred here unless the factory is intended to retain ownership.  The unique_ptr return value can be assigned to a shared_ptr.